### PR TITLE
Add type hints to `Image.py`

### PR DESCRIPTION
**Pull request by hugovk**
Still some to do later.

### CODE REVIEW

File: src/PIL/Image.py [hunk 1]
```diff
@@ -120,7 +120,7 @@ class DecompressionBombError(Exception):
     cffi = None
 
 
-def isImageType(t):
+def isImageType(t: Image) -> bool:
```
Comment (nulano) [inline src/PIL/Image.py:123]:
```suggestion
def isImageType(t: Any) -> TypeGuard[Image]:
```

File: src/PIL/Image.py [hunk 2]
```diff
@@ -1117,11 +1122,11 @@ def convert_transparency(m, v):
...
-        colors=256,
-        method=None,
-        kmeans=0,
+        colors: int = 256,
+        method: Quantize | None = None,
+        kmeans: int = 0,
         palette=None,
-        dither=Dither.FLOYDSTEINBERG,
+        dither: Dither = Dither.FLOYDSTEINBERG,
     ):
```
Comment (nulano) [inline src/PIL/Image.py:1130]:
```suggestion
    ) -> Image:
```

File: src/PIL/Image.py [hunk 3]
```diff
@@ -1364,7 +1370,7 @@ def getcolors(self, maxcolors=256):
             return out
         return self.im.getcolors(maxcolors)
 
-    def getdata(self, band=None):
+    def getdata(self, band: int | None = None) -> Image:
```
Comment (radarhere) [inline src/PIL/Image.py:1373]: This returns an ImagingCore instance, not Image.

File: src/PIL/Image.py [hunk 4]
```diff
@@ -1117,12 +1123,12 @@ def convert_transparency(m, v):
 
     def quantize(
         self,
-        colors=256,
-        method=None,
-        kmeans=0,
+        colors: int = 256,
+        method: Quantize | None = None,
+        kmeans: int = 0,
         palette=None,
```
Comment (radarhere) [inline src/PIL/Image.py:1129]:
```suggestion
        palette: Image | None = None,
```

File: src/PIL/Image.py [hunk 5]
```diff
@@ -508,22 +509,22 @@ def __init__(self):
...
-    def size(self):
+    def size(self) -> tuple[int, int]:
         return self._size
 
     @property
     def mode(self):
         return self._mode
 
-    def _new(self, im):
+    def _new(self, im: Image) -> Image:
```
Comment (radarhere) [inline src/PIL/Image.py:527]: `im` is an ImagingCore instance, not Image.

File: src/PIL/Image.py [hunk 6]
```diff
@@ -1341,7 +1347,7 @@ def getbbox(self, *, alpha_only: bool = True) -> tuple[int, int, int, int]:
         self.load()
         return self.im.getbbox(alpha_only)
 
-    def getcolors(self, maxcolors=256):
+    def getcolors(self, maxcolors: int = 256) -> list[tuple[int, int]] | None:
```
Comment (radarhere) [inline src/PIL/Image.py:1350]:
```suggestion
    def getcolors(
        self, maxcolors: int = 256
    ) -> list[tuple[int, int | tuple[int, ...]]] | None:
```
```pycon
>>> from PIL import Image
>>> im = Image.new("RGB", (1, 1))
>>> im.getcolors()
[(1, (0, 0, 0))]
```

File: src/PIL/Image.py [hunk 7]
```diff
@@ -1387,7 +1393,7 @@ def getdata(self, band=None):
             return self.im.getband(band)
         return self.im  # could be abused
 
-    def getextrema(self):
+    def getextrema(self) -> tuple[float, float] | tuple[tuple[float, float], ...]:
```
Comment (radarhere) [inline src/PIL/Image.py:1396]:
```suggestion
    def getextrema(self) -> tuple[float, float] | tuple[tuple[int, int], ...]:
```
We don't have multi-band images with float values.

File: src/PIL/Image.py (hunk 5 above)
Comment (hugovk) [reply src/PIL/Image.py:527]: We can deal with the `ImagingCore` type later.
```suggestion
    def _new(self, im: Image):
```

File: src/PIL/Image.py (hunk 3 above)
Comment (hugovk) [reply src/PIL/Image.py:1373]:
```suggestion
    def getdata(self, band: int | None = None):
```

File: src/PIL/Image.py [hunk 8]
```diff
@@ -508,22 +509,22 @@ def __init__(self):
...
-    def size(self):
+    def size(self) -> tuple[int, int]:
         return self._size
 
     @property
     def mode(self):
         return self._mode
 
-    def _new(self, im):
+    def _new(self, im: Image):
```
Comment (radarhere) [inline src/PIL/Image.py:527]:
```suggestion
    def _new(self, im) -> Image:
```
This was corrected the wrong way around - it returns an `Image`, but `im` is not.

File: src/PIL/Image.py [hunk 9]
```diff
@@ -1341,7 +1347,9 @@ def getbbox(self, *, alpha_only: bool = True) -> tuple[int, int, int, int]:
         self.load()
         return self.im.getbbox(alpha_only)
 
-    def getcolors(self, maxcolors=256):
+    def getcolors(
+        self, maxcolors: int = 256
+    ) -> list[tuple[int, int | tuple[int, ...]]] | None:
```
Comment (hugovk) [inline src/PIL/Image.py:1352]: Let's deal with this after this release:
```
mypy: commands[0]> mypy src
src/PIL/Image.py:1372: error: Incompatible return value type (got
"List[Tuple[Any, int]]", expected
"Optional[List[Tuple[int, Union[int, Tuple[int, ...]]]]]")  [return-value]
                return out
                       ^~~
Found 1 error in 1 file (checked 102 source files)
mypy: exit 1 ([16](https://github.com/python-pillow/Pillow/actions/runs/8504209278/job/23290792512?pr=7924#step:8:17).84 seconds) /home/runner/work/Pillow/Pillow> mypy src pid=2141
```


```suggestion
    def getcolors(self, maxcolors: int = 256):
```

**Review by radarhere — APPROVED** [review]

**Comment by Yay295** [comment]
I think `ImagingCore` would be typed as `typing.Any`, pending a more concrete definition. Though I suppose leaving it untyped for now works as well.

**Pull request merged**